pinot-plugins/pinot-stream-ingestion/pinot-kafka-2.0/src/main/java/org/apache/pinot/plugin/stream/kafka20/metadata/KafkaMessageMetadataFactory.java

```diff
@@ -0,0 +1,29 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.plugin.stream.kafka20.metadata;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+
+
+public class KafkaMessageMetadataFactory implements MessageMetadataFactory {
+  @Override
+  public KafkaMessageMetadata create(ConsumerRecord<?, ?> consumerRecord) {
+    return new KafkaMessageMetadata(consumerRecord.timestamp());
+  }
+}
```

Comment:
This is a little bit confusing as the interface MessageMetadataFactory is specific to kafka (ConsumerRecord argument of the method belongs to kafka) and then the class KafkaMessageMetadataFactory implements that!
Since the logic is simple, I would suggest consolidating three classes `MessageMetadataFactory`, `KafkaMessageMetadataFactory`, and `NoopMessageMetadataFactory` into one. Something like the following (I used extractor suffix instead of factory):
```java
public interface KafkaMessageMetadataExtractor {

  KafkaMessageMetadata extract(ConsumerRecord<?, ?> consumerRecord);

  static KafkaMessageMetadataExtractor build(boolean populateMetadata) {
    if (populateMetadata) {
      return record -> new KafkaMessageMetadata(record.timestamp());
    } else {
      return record -> null;
    }
  }
}
```

Reply:
+1 to the above suggestion. 
You are processing a record and not really creating an new instance of anything. So the use of factory pattern is confusing.

Reply:
Thanks @sajjad-moradi @navina for your feedback. One on the reason why I've implemented it like this is precisely to avoid this conditional because this config driven and the evaluation result will be the same for every record this would be inefficient. The other reason is right now the metadata includes only 1 property but more may be added later and the construction of `KafkaMessageMetadata` may become more complex. Would it help if I rename the classes to remove the word Factory? I can replace with Extractor or Creator or Provider

Reply:
Oh, didn't pay attention to the example, you are actually returning a lambda. Yes, in this case you are right

Reply:
Change made